sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java

```diff
@@ -482,40 +506,62 @@ private void captureSessionToken(RxDocumentServiceRequest request, Map<String, S
-            Strings.areEqual(requestConsistencyLevel, ConsistencyLevel.SESSION.toString()) ||
-                (this.defaultConsistencyLevel == ConsistencyLevel.SESSION &&
-                    // skip applying the session token when Eventual Consistency is explicitly requested
-                    // on request-level for data plane operations.
-                    // The session token is ignored on teh backend/gateway in this case anyway
-                    // and the session token can be rather large (even run in the 16 KB header length problem
-                    // on the gateway - so not worth sending when not needed
-                    (!request.isReadOnlyRequest() ||
-                        request.getResourceType() != ResourceType.Document ||
-                        !Strings.areEqual(requestConsistencyLevel, ConsistencyLevel.EVENTUAL.toString())));
-
-        if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
-            if (!sessionTokenApplicable || isMasterOperation(request.getResourceType(), request.getOperationType())) {
+        // Master resource operations don't require session token.
+        if (isMasterOperation(request.getResourceType(), request.getOperationType())) {
+            if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
                 request.getHeaders().remove(HttpConstants.HttpHeaders.SESSION_TOKEN);
             }
-            return; //User is explicitly controlling the session.
+            return Mono.empty();
         }
 
-        if (!sessionTokenApplicable || isMasterOperation(request.getResourceType(), request.getOperationType())) {
-            return; // Only apply the session token in case of session consistency and when resource is not a master resource
+        if (!Strings.isNullOrEmpty(request.getHeaders().get(HttpConstants.HttpHeaders.SESSION_TOKEN))) {
+            return Mono.empty(); // User is explicitly controlling the session.
         }
 
-        //Apply the ambient session.
-        String sessionToken = this.sessionContainer.resolveGlobalSessionToken(request);
+        boolean sessionConsistency = RequestHelper.getConsistencyLevelToUse(this.gatewayServiceConfigurationReader,
+            request) == ConsistencyLevel.SESSION;
 
-        if (!Strings.isNullOrEmpty(sessionToken)) {
-            headers.put(HttpConstants.HttpHeaders.SESSION_TOKEN, sessionToken);
+        if (!sessionConsistency ||
+            (!request.isReadOnlyRequest() && request.getOperationType() != OperationType.Batch && !this.useMultipleWriteLocations)) {
+            return Mono.empty();
+            // Only apply the session token in case of session consistency and if request is read only,
+            // apply token for write request only if batch operation or multi master
+        }
+
+        String partitionKeyRangeId = request.getHeaders().get(HttpConstants.HttpHeaders.PARTITION_KEY_RANGE_ID);
```

Comment:
what happens if feedRange is configured in the query? i.e., `QueryFeedOptions.setFeedRange()`
if FeeRange is set then is it guaranteed this code work?

Could you please add a test for this scenario?

Reply:
Added test for query via feedRange